EngineLayer/ModernSearch/ModernSearchEngine.cs

```diff
@@ -153,12 +161,38 @@ protected override MetaMorpheusEngineResults RunSpecific()
 
         #region Private Methods
 
-        private void CalculatePeptideScores(IMsDataScan<IMzSpectrum<IMzPeak>> spectrum, double[] peptideScores)
+        private void CalculatePeptideScores(IMsDataScan<IMzSpectrum<IMzPeak>> spectrum, double[] peptideScores, double thePrecursorMass)
         {
+            List<IMzPeak> experimentalPeaks = new List<IMzPeak>();
             for (int i = 0; i < spectrum.MassSpectrum.Size; i++)
             {
-                var theAdd = 1 + spectrum.MassSpectrum[i].Intensity / spectrum.TotalIonCurrent;
-                var experimentalPeakInDaltons = spectrum.MassSpectrum[i].Mz - Constants.protonMass;
+                experimentalPeaks.Add(spectrum.MassSpectrum[i]);
+            }
+            if (addCompIons)
+            {
+                //If HCD
+                if (lp.Contains(ProductType.B) | lp.Contains(ProductType.Y))
+                {
+                    for (int i = 0; i < spectrum.MassSpectrum.Size; i++)
+                    {
+                        experimentalPeaks.Add(new MzPeak((thePrecursorMass - spectrum.MassSpectrum[i].Mz + Constants.protonMass * 2), (spectrum.MassSpectrum[i].Intensity / 100)));
+                    }
+                }
+                //If ETD
+                if (lp.Contains(ProductType.C) | lp.Contains(ProductType.Zdot))
+                {
+                    for (int i = 0; i < spectrum.MassSpectrum.Size; i++)
+                    {
+                        experimentalPeaks.Add(new MzPeak((thePrecursorMass - spectrum.MassSpectrum[i].Mz + Constants.protonMass * 3), (spectrum.MassSpectrum[i].Intensity / 100)));
+                    }
+                }
+            }
+            IEnumerable<IMzPeak> sortedPeaksMZ = experimentalPeaks.OrderBy(x => x.Mz);
```

Comment:
seems a bit inefficient the way it is done here, with creation of peak objects.. can you reorganize this so that if the compIons is turned off, the matching is as efficient as before?